### ISSUE TITLE
Library Question: More Generic Usage?

### ISSUE THREAD

**Issue by mergesort** [kudrykv/alfred-craftdocs-searchindex#8]
Hi there,

I've been a long-time Alfred user and really, really love the Craft Docs Search integration you built. A few months ago I switched to using Raycast and have found that this extension is what I miss most. I don't want this to come off as me asking you to build a [Raycast extension](https://www.raycast.com/store) because I appreciate the hard work and effort that went into building and maintaining the Alfred extension, and am guessing Alfred is your preferred launcher given the love you put into this integration.

I tried to play around a bit with the executable that's bundled to see if I could craft a Raycast version of my own and was unsuccessful, so I wanted to ask if there's any form of documentation, guidance, or internal API for the plugin that you can provide that would make that easier to repurpose the work you've done for another launcher?

Thank you so much for all the hard work!

**Comment by kudrykv** [comment]
Hi @mergesort!

Haha, I've moved to Raycast some time ago and enjoying it :)
I've moved away from Craft Docs, so I didn't build one for Raycast.

If you want to take a bite yourself, here are a few SQLite configuration references you'll need:
- use [`sql.js`](https://github.com/kudrykv/raycast-safari-history/blob/1f189ebbcddd07d807fa597ff185d6a0b3b89323/src/useHistorySearch.ts#L2);
- [load DB with a wasm binary file](https://github.com/kudrykv/raycast-safari-history/blob/1f189ebbcddd07d807fa597ff185d6a0b3b89323/src/useHistorySearch.ts#L45-L47);
- [here is the wasm file itself](https://github.com/kudrykv/raycast-safari-history/blob/main/assets/sql-wasm.wasm).

Raycast doesn't expose convenient ways of interacting with SQLite ¯\\\_(ツ)\_\/¯

Or you can leave it to me, and I'll possibly make it in a few days :) You choose!

**Comment by mergesort** [comment]
I didn't want to impose but if you're offering I'd gladly take you up on that @kudrykv. 🙂

**Comment by DFelten** [comment]
I am also interested in a Raycast expansion for Craft :)

**Comment by kudrykv** [comment]
So you can try it out if you know how to run the development mode :) https://github.com/kudrykv/extensions/tree/craftdocs/extensions/craftdocs

In the meantime, I'm polishing it and prepping for the store

**Comment by DFelten** [comment]
Thanks, it's working great :) I'm looking forward for the polishing. Craft icon for example is missing. Is it possible to show the first line of the article or some other preview?

**Comment by kudrykv** [comment]
@DFelten do you mean finding documents instead (rather than blocks right now), and show their previews?

**Comment by kudrykv** [comment]
I envision it to be as it is currently, roughly this form:
<img width="862" alt="CleanShot 2022-04-22 at 12 33 45@2x" src="https://user-images.githubusercontent.com/12697803/164679326-44193c6e-3acf-46db-a1f8-cf98dc995f58.png">

The index, where the data comes from, is somewhat obfuscated, hence showing beautiful preview isn't possible.

**Comment by DFelten** [comment]
I think it makes sense to display the documents first, and then the blocks. The Alfred extension indicated whether the content was a block or a document. 

The example looks great :) A small preview of the content (as far as possible) is definitely helpful.

**Comment by kudrykv** [comment]
@DFelten it's the best what I can get from it

https://user-images.githubusercontent.com/12697803/164712026-259534d6-e85d-4835-b868-5c15132082ee.mp4

**Comment by DFelten** [comment]
This looks really good, thank you very much! This preview helps enormously.

**Comment by kudrykv** [comment]
@DFelten pushed changes — https://github.com/kudrykv/extensions/tree/craftdocs/extensions/craftdocs
You can build it for yourself for now

**Comment by DFelten** [comment]
Thanks :) The information seems to be shown in the wrong order. The last line of the document appears at the top.

The preview is not very pretty, but it is informative. You also don't see if they are sections, do you? That would at least provide some structure.

Is the preview also possible for the block search?

**Comment by kudrykv** [comment]
I have this data:
- id — block id;
- content — data in the block (text);
- type — block type (text, image, url, file, code);
- entityType — what given block is (page, block, document);
- customRank — Craft search ranking;
- isTodo — whether it's a todo or not (no spec whether it is marked as completed);
- documentId — id of the document to which the document belongs;
- stamp — yet another ID;
- exactMatchContent — another internal gut of Craft; either empty or content with no spaces.

I have no access to block order in the document whatsoever

**Comment by DFelten** [comment]
I thought you would get the contents as an array or list and could reverse the order. 

Then I thing this preview is fine. But I would add this preview as an optional option to the settings of the extension.

**Comment by kudrykv** [comment]
I'll try reversing the order to see whether that helps anything. There is a slight chance the order would roughly match what's actually in the document, but the chances of that are slim ¯\\\_(ツ)\_\/¯

**Comment by DFelten** [comment]
If you've changed that, I'd be happy to check it out with a few examples.

It might look a little better without the dividers.

**Comment by kudrykv** [comment]
@DFelten natural order (what DB returns) is also a mess. Reverse or no reverse wouldn't help. Here's an example on my FFMpeg doc:
![CleanShot 2022-04-22 at 16 54 28@2x](https://user-images.githubusercontent.com/12697803/164728927-54b8cc40-c48f-4f39-8c7c-8db0901f4c22.png)

The only option I see here is to connect to the Realm DB. Last time it was a huge problem, so I abandoned it.
Unless something has changed and it's possible to connect to their original DB without too much hassle, the response would be as that.

**Comment by kudrykv** [comment]
Okay, Realm may be easier now. I'll investigate this option first.

**Comment by kudrykv** [comment]
Realm is still a problem =/
This needs to be a separate effort.

**Comment by DFelten** [comment]
Then I would say that the preview should be optionally activatable, if you still want to include it despite the problems. 

Personally, I still find it practical. Maybe this will get better with further Craft versions.

**Comment by kudrykv** [comment]
So here is what I have and what'll be roughly there (it's in my repo already):
- a tick to enable a detailed view;
- the list of blocks;
- the detailed list of documents.

![CleanShot 2022-04-22 at 20 38 38](https://user-images.githubusercontent.com/12697803/164766176-0c1091a7-0a8d-43e5-9e35-19d278515633.png)

https://user-images.githubusercontent.com/12697803/164766351-94149957-35f3-418f-a716-8fa964788e41.mp4

https://user-images.githubusercontent.com/12697803/164766361-14a3724b-e057-4d37-998c-e4bd9c27c209.mp4

**Comment by DFelten** [comment]
Thank you very much, the extension works very well.

**Comment by kudrykv** [comment]
@DFelten, out of curiosity, do you use it with the list or a detailed view?

**Comment by DFelten** [comment]
Currently I prefer the detailed view. I have some smaller notes and can see them so well. 

I have a lot of code snippets. You wrote that you can recognize the type of them. Would it be good possible to design them differently?

**Comment by kudrykv** [comment]
@DFelten makes sense! Be warned, though, that those snippets will be broken (as it is now).

**Comment by kudrykv** [comment]
@DFelten please check now (pushed).

**Comment by DFelten** [comment]
This looks great, thanks!

![CleanShot 2022-04-28 at 10 13 46](https://user-images.githubusercontent.com/1169185/165708344-34b4e4e4-e497-45bb-9d8d-83ee013a1f6e.png)

An other suggestion: Is it possible to add a command to create a page with Craft from Raycast?

**Comment by kudrykv** [comment]
@DFelten it should already work
![CleanShot 2022-04-28 at 14 55 40](https://user-images.githubusercontent.com/12697803/165746701-dba62184-4622-43dd-a89f-d52d0b9f2359.png)

**Comment by kudrykv** [comment]
@mergesort were you able to install it locally while it's not in the store yet?

**Comment by DFelten** [comment]
> @DFelten it should already work
> ![CleanShot 2022-04-28 at 14 55 40](https://user-images.githubusercontent.com/12697803/165746701-dba62184-4622-43dd-a89f-d52d0b9f2359.png)

Yes, but only when nothing is found. For example when I would like to create a page named "Magic", but "Magic Monday" already exists. Then it's not possible to create that page via Raycast. A custom command for creating a page could be useful.

**Comment by kudrykv** [comment]
@DFelten, oh, gotcha.

I've added the last item to always be "create document" if you search for something. So if you have lots of results, you can quickly navigate to the last item with cmd+down and see your search results: maybe, you already have what you need.

![CleanShot 2022-04-29 at 14 03 04](https://user-images.githubusercontent.com/12697803/165932666-1eb394a5-bf36-4441-91b2-cae085a188f8.png)

The change is in the repo

**Comment by mergesort** [comment]
So sorry I didn't see this, I've been on vacation for the last week. I installed it and it works super well as is! I switched to the Details view and prefer it as well, I think that's what I'll use going forward since I'm almost always using it as a doc launcher and like having a cleaner list of all the docs available for me to open.

Thank you so much again for porting this over! Can't wait to see it on the Extensions store. 🙂

**Comment by DFelten** [comment]
> @DFelten, oh, gotcha.
> 
> I've added the last item to always be "create document" if you search for something. So if you have lots of results, you can quickly navigate to the last item with cmd+down and see your search results: maybe, you already have what you need.
> 
> ![CleanShot 2022-04-29 at 14 03 04](https://user-images.githubusercontent.com/12697803/165932666-1eb394a5-bf36-4441-91b2-cae085a188f8.png)
> 
> The change is in the repo

Thanks, thats working :) This has become a great extension!

**Comment by kudrykv** [comment]
@DFelten @mergesort it's in the store :) https://github.com/raycast/extensions/pull/1595
<img width="878" alt="CleanShot 2022-05-23 at 11 03 06@2x" src="https://user-images.githubusercontent.com/12697803/169772578-bdac59f8-6217-4d01-a5de-8523d8e77e16.png">

**Comment by DFelten** [comment]
Thanks :) Great extension!

**Comment by mergesort** [comment]
Thanks a lot, upgraded to the store version!

**Comment by ArttuLai** [comment]
Hey great extension, thanks for doing this even though you’re not using Craft anymore.

I have tiny low priority request though: would it be possible to replace the emoji icons with something a bit more clean like SF Symbols? If you’re familiar with SF Symbols, I was thinking something like `doc.richtext` for documents and `doc.on.doc` for subpages/blocks.

I also find hard to distinguish 📄 and 📑 when they’re displayed in the small text size Raycast uses. Also the extension is so great I think it deserves nice icons and not just emojis. I can help however I can, help designing the icons etc, but I’m not dev-enough to be able to just do a PR on my own to the Raycast extensions repo (also I don’t think it’s for me to decide what icons you want to use).

<img width="121" alt="image" src="https://user-images.githubusercontent.com/42007282/178021556-e1360249-ab17-4740-9618-8ca2de780cf6.png">
<img width="121" alt="image" src="https://user-images.githubusercontent.com/42007282/178020506-d8ddb5ec-43ad-41e0-b458-a2b62dd1faad.png">

**Comment by kudrykv** [comment]
hi @ArttuLai!

This is a problem indeed. Thank you for looking into icons! I'll pick the ones you've suggested.
I didn't know about SF Symbols before. Raycast doesn't allow me to refer to them directly, but I'd be able to export them to something more conventional.

**Comment by kudrykv** [comment]
@ArttuLai, icons would look smth like this, then
![CleanShot 2022-07-08 at 19 42 47](https://user-images.githubusercontent.com/12697803/178034964-9aa56f68-2a16-4aba-aee8-7d3fef6b8048.png)

**Comment by ArttuLai** [comment]
Thank you. Yes it looks great.

I think you need separate icons for light and dark mode though, or tint it using `tintColor: Color.PrimaryText` for example.

**Comment by kudrykv** [comment]
Yup
![CleanShot 2022-07-08 at 19 46 26](https://user-images.githubusercontent.com/12697803/178035434-ebffe674-7fbf-4181-bdaf-ab446d1709db.png)

**Comment by ArttuLai** [comment]
Beautiful 😊

**Comment by kudrykv** [comment]
https://github.com/raycast/extensions/pull/2215

**Comment by youtextme** [comment]
Nice — that Alfred → Raycast jump is common once a missing workflow lands. If you’re still mapping the rest of the stack: free Raycast already covers clipboard history + snippets + most Store extensions; only pay Pro if a weekly job actually needs it.

https://www.raycast.com/?via=girish
Affiliate — I earn a commission if you upgrade to Pro.
